### PR TITLE
checking that a message is old or invalid is now under one lock

### PR DESCRIPTION
**Pull request by zalmen**
Fixes: https://github.com/spacemeshos/go-spacemesh/issues/449

### CODE REVIEW

File: p2p/gossip/protocol.go [hunk 1]
```diff
@@ -138,18 +141,23 @@ func (prot *Protocol) Close() {
...
+// markMessage adds the calcHash to the old message queue so the message won't be processed in case received again.
+// Returns true if message was already processed before
+func (prot *Protocol) markMessage(h hash) bool {
 	prot.oldMessageMu.Lock()
-	prot.oldMessageQ[h] = struct{}{}
+	var ok bool
+	if _, ok = prot.oldMessageQ[h]; !ok {
+		prot.oldMessageQ[h] = struct{}{}
+	}
+	prot.Log.Debug("marking message as old, hash %v, is already old %v", h, ok)
```
Comment (gavraz) [inline p2p/gossip/protocol.go:155]: The message is confusing, you are not marking the message as old here.

File: p2p/gossip/protocol.go [hunk 2]
```diff
@@ -138,18 +141,23 @@ func (prot *Protocol) Close() {
...
 	prot.oldMessageMu.Unlock()
+	return ok
```
Comment (gavraz) [inline p2p/gossip/protocol.go:157]: Personally, I prefer clarity over a single line code duplication. I suggest you avoid the ok var, have two return statements and "duplicate" the unlock.

File: p2p/gossip/protocol.go [hunk 3]
```diff
@@ -237,21 +248,25 @@ func (prot *Protocol) removePeer(peer crypto.PublicKey) {
 	prot.peersMutex.Unlock()
 }
 
-func (prot *Protocol) isOldMessage(h hash) bool {
-	var oldmessage bool
+// marks a hash as old message and check message validity
+func (prot *Protocol) markAndValidateMessage(h hash, msg *pb.ProtocolMessage) (isOldMessage, isInvalid bool) {
 	prot.oldMessageMu.RLock()
```
Comment (gavraz) [inline p2p/gossip/protocol.go]: Change the read lock to write lock.

File: p2p/gossip/protocol.go [hunk 4]
```diff
@@ -271,38 +286,38 @@ func (prot *Protocol) processMessage(h hash, msg *pb.ProtocolMessage) error {
...
-			return err
-		}
-
-		err = prot.processMessage(hash, msg)
+	} else {
+		//todo - processMessage is non-blocking, need to check validation and not propagate invalid messages
+		err = prot.processMessage(msg)
 		if err != nil {
-			return err
+			return
```
Comment (gavraz) [inline p2p/gossip/protocol.go:318]: Check if you want to log on that return.

**Review by gavraz — CHANGES_REQUESTED** [review]

File: p2p/gossip/protocol.go (hunk 1 above)
Comment (zalmen) [reply p2p/gossip/protocol.go:155]: The "is already old" tell you whether it was already old

File: p2p/gossip/protocol.go (hunk 3 above)
Comment (zalmen) [reply p2p/gossip/protocol.go]: fixed

File: p2p/gossip/protocol.go (hunk 4 above)
Comment (zalmen) [reply p2p/gossip/protocol.go:318]: This is intentionally, in the next PR I'll change the way we pass messages to protocol handlers

**Review by gavraz — APPROVED** [review]

**Pull request merged**